Version 2.3

qwales1 opened the pull request:
This resolves #134, #145, and #147 

1. Adds the ability to have multiple instances of re-base instead re-base being a singleton (#134)
2. Adds support for having 'named' firebase apps, previously re-base could only use the default app ( #134 )
3. Adds `base.initializedApp` property that provides the initialized firebase app for that instance of re-base.
4. Adds a `base.remove` method which allows you to remove data the specified location ( #145 ) 
5. Refactors out all the database, auth, user, and utility methods into separate functions so the internals can be tested (still need to add the tests 😬 )
6. Refactors out all the tests into separate files
7. Reduces the amount of state shared between tests
8. Updates travis to run tests on the latest Firefox extended support release
9. Removes dependency on es6-symbol
10. Exposes new firebase Messaging service
11. Updates all the devDependencies to the latest versions. 
Notably Babel 5 to Babel 6, Karma 0.13 to Karma 1.3, React 0.13 to React 15.3

tylermcginnis commented:
Nice! What are the specific breaking changes? (And do we have an upgrade path in place?)

qwales1 commented:
@tylermcginnis Definitely wanted your opinion/advice on this because there is a way to do this without the breaking change (which after writing this out I actually prefer). 

Since 2.0, `re-base` exposed the firebase services `firebase.database`, `firebase.app`, `firebase.auth`, and `firebase.storage`.  `re-base` was a singleton and it could only initialize one firebase app ( named `'[DEFAULT]'`). If you called `base.auth()` you would get back the auth service for the `'[DEFAULT]'` app that re-base initialized. That is the behavior I would expect. 

This PR allows you to initialize multiple firebase apps with different configs, a feature that firebase 3.x introduced, and exposing the firebase services directly could get a little weird. Here is an example of what _could_ happen without the breaking change. 

```javascript
//initializing a firebase app instance with the name 'myBase'
var myBase = Rebase.createClass(configObject, 'myBase');
//in some component elsewhere we want to use the firebase database service directly
var ref = myBase.database().ref(); // <-- THROWS ERROR 
//No Firebase App '[DEFAULT]' has been created - call Firebase App.initializeApp().(…)

```
This happens because re-base was exposing `firebase.database` instead of the initialized app's database service with is what this PR does instead. With this PR : 

```javascript 
var myBase = Rebase.createClass(configObject, 'myBase');
var ref = myBase.database().ref(); // <-- WORKS
```

The alternative, non-breaking change would be to still expose `firebase.database`, `firebase.storage` etc. And then would need to document that, if you are using multiple firebase apps, you need to make sure you provide the firebase services you want to use with the **correct** firebase app which re-base could expose for convenience. 

```javascript
var myBase = Rebase.createClass(configObject, 'myBase');
myBase.database(myBase.initializedApp).ref(); <-- WORKS
myBase.database().ref(); // <-- THROWS ERROR in this case.
// This shouldn't break anyones app though. You can currently only 
// initialize the '[DEFAULT]' app with rebase.createClass so this can't actually happen
// until people start to use multiple firebase apps

```
The other part that could have been breaking is that I removed the dependency on `es6-symbol`. Travis was running Firefox 31 (no Symbol support), which I updated to run the latest extended support release. There was a `for..of` loop that babel was transforming that introduced Symbol. I just re-factored that out so it won't be an issue. No need for Symbol support anymore. 

So if you are ok with the third option above I can re-work this and call it 2.3 🎉  . Let me know what you think.

qwales1 commented:
@tylermcginnis I updated this PR to be non-breaking and edited the list of changes above.
